src/DotNetLightning.Core/Utils/Keys.fs

```diff
@@ -0,0 +1,348 @@
+    | HtlcBasepoint of PubKey
+    with
+    member this.RawPubKey(): PubKey =
+        let (HtlcBasepoint pubKey) = this
+        pubKey
+
+    member this.ToBytes(): array<byte> =
+        this.RawPubKey().ToBytes()
+
+    override this.ToString() =
+        this.RawPubKey().ToString()
+
+type HtlcBasepointSecret =
+    | HtlcBasepointSecret of Key
+    with
+    member this.RawKey(): Key =
+        let (HtlcBasepointSecret key) = this
+        key
+
+    member this.HtlcBasepoint(): HtlcBasepoint =
+        HtlcBasepoint(this.RawKey().PubKey)
+
+type HtlcPubKey =
+    | HtlcPubKey of PubKey
+    with
+    member this.RawPubKey(): PubKey =
+        let (HtlcPubKey pubKey) = this
+        pubKey
+
+    member this.ToBytes(): array<byte> =
+        this.RawPubKey().ToBytes()
+
+    override this.ToString() =
+        this.RawPubKey().ToString()
+
+type HtlcPrivKey =
+    | HtlcPrivKey of Key
+    with
+    member this.RawKey(): Key =
+        let (HtlcPrivKey key) = this
+        key
+
+    member this.HtlcPubKey(): HtlcPubKey =
+        HtlcPubKey <| this.RawKey().PubKey
+
+type NodeSecret =
+    | NodeSecret of Key
+    with
+    member this.RawKey(): Key =
+        let (NodeSecret key) = this
+        key
+
+    member this.NodeId(): NodeId =
+        NodeId (this.RawKey().PubKey)
+
+/// In usual operation we should not hold secrets on memory. So only hold pubkey
+type ChannelPubKeys = {
+    FundingPubKey: FundingPubKey
+    RevocationBasepoint: RevocationBasepoint
+    PaymentBasepoint: PaymentBasepoint
+    DelayedPaymentBasepoint: DelayedPaymentBasepoint
+    HtlcBasepoint: HtlcBasepoint
+}
+
+type CommitmentPubKeys = {
+    RevocationPubKey: RevocationPubKey
+    PaymentPubKey: PaymentPubKey
+    DelayedPaymentPubKey: DelayedPaymentPubKey
+    HtlcPubKey: HtlcPubKey
+}
+
+type PerCommitmentPoint =
+    | PerCommitmentPoint of PubKey
+    with
+    member this.RawPubKey(): PubKey =
+        let (PerCommitmentPoint pubKey) = this
+        pubKey
+
+    static member BytesLength: int = PubKey.BytesLength
+
+    static member FromBytes(bytes: array<byte>): PerCommitmentPoint =
+        PerCommitmentPoint <| PubKey bytes
+
+    member this.ToBytes(): array<byte> =
+        this.RawPubKey().ToBytes()
+
+type [<Struct>] CommitmentNumber =
+    | CommitmentNumber of UInt48
+    with
+    member this.Index() =
+        let (CommitmentNumber index) = this
+        index
+
+    override this.ToString() =
+        sprintf "%012x (#%i)" (this.Index().UInt64) (UInt48.MaxValue - this.Index()).UInt64
+
+    static member LastCommitment: CommitmentNumber =
+        CommitmentNumber UInt48.Zero
+
+    static member FirstCommitment: CommitmentNumber =
+        CommitmentNumber UInt48.MaxValue
+
+    member this.PreviousCommitment(): CommitmentNumber =
+        CommitmentNumber(this.Index() + UInt48.One)
+
+    member this.NextCommitment(): CommitmentNumber =
+        CommitmentNumber(this.Index() - UInt48.One)
+
+and [<Struct>] ObscuredCommitmentNumber =
```

Comment:
@canndrew does this really need an `and`? doesn't look like

Reply:
Fixed.